src/cargo/util/lints.rs

```diff
@@ -0,0 +1,226 @@
+    }
+}
+
+impl From<TomlLintLevel> for LintLevel {
+    fn from(toml_lint_level: TomlLintLevel) -> LintLevel {
+        match toml_lint_level {
+            TomlLintLevel::Allow => LintLevel::Allow,
+            TomlLintLevel::Warn => LintLevel::Warn,
+            TomlLintLevel::Deny => LintLevel::Deny,
+            TomlLintLevel::Forbid => LintLevel::Forbid,
+        }
+    }
+}
+
+/// By default, cargo will treat any optional dependency as a [feature]. As of
+/// cargo 1.60, these can be disabled by declaring a feature that activates the
+/// optional dependency as `dep:<name>` (see [RFC #3143]).
+///
+/// In the 2024 edition, `cargo` will stop exposing optional dependencies as
+/// features implicitly, requiring users to add `foo = ["dep:foo"]` if they
+/// still want it exposed.
+///
+/// For more information, see [RFC #3491]
+///
+/// [feature]: https://doc.rust-lang.org/cargo/reference/features.html
+/// [RFC #3143]: https://rust-lang.github.io/rfcs/3143-cargo-weak-namespaced-features.html
+/// [RFC #3491]: https://rust-lang.github.io/rfcs/3491-remove-implicit-features.html
+const IMPLICIT_FEATURES: Lint = Lint {
+    name: "implicit-features",
+    desc: "warn about the use of unstable features",
+    groups: &[RUST_2024_COMPATIBILITY],
+    default_level: LintLevel::Allow,
+    edition_lint_opts: Some((Edition::Edition2024, LintLevel::Deny)),
+};
+
+pub fn check_implicit_features(
+    pkg: &Package,
+    path: &Path,
+    lints: &TomlToolLints,
+    error_count: &mut usize,
+    gctx: &GlobalContext,
+) -> CargoResult<()> {
+    let lint_level = IMPLICIT_FEATURES.level(lints, pkg.manifest().edition());
+    if lint_level == LintLevel::Allow {
+        return Ok(());
+    }
+
+    let manifest = pkg.manifest();
+    let user_defined_features = manifest.resolved_toml().features();
+    let features = user_defined_features.map_or(HashSet::new(), |f| {
+        f.keys().map(|k| InternedString::new(&k)).collect()
+    });
+    // Add implicit features for optional dependencies if they weren't
+    // explicitly listed anywhere.
+    let explicitly_listed = user_defined_features.map_or(HashSet::new(), |f| {
+        f.values()
+            .flatten()
+            .filter_map(|v| match FeatureValue::new(v.into()) {
+                Dep { dep_name } => Some(dep_name),
+                _ => None,
+            })
+            .collect()
+    });
+
+    for dep in manifest.dependencies() {
+        let dep_name_in_toml = dep.name_in_toml();
+        if !dep.is_optional()
+            || features.contains(&dep_name_in_toml)
+            || explicitly_listed.contains(&dep_name_in_toml)
+        {
+            continue;
+        }
+        if lint_level == LintLevel::Forbid || lint_level == LintLevel::Deny {
+            *error_count += 1;
+        }
+        let level = lint_level.to_diagnostic_level();
+        let manifest_path = rel_cwd_manifest_path(path, gctx);
+        let message = level.title("unused optional dependency").snippet(
```

Comment:
Should we have a transitional note that implicit features are no longer supported?